### PR TITLE
[WIP] Chuigda Generated Parser

### PR DESCRIPTION
**Pull request by chuigda**
It will be funny.

### CODE REVIEW

File: src/syntax/concrete.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+pub trait TypeVisitor {
+    type TypeResult;
+}
+
+pub trait StmtVisitor {
+    type StmtResult;
+}
+
+pub trait ConcreteVisitor : DeclVisitor + ExprVisitor + TypeVisitor + StmtVisitor {
```
Comment (chuigda) [inline src/syntax/concrete.rs]: @HoshinoTented Do you like such Visitor API? Or you prefer something else?

**Comment by chuigda** [comment]
Okay I've gave up such Visitor API since it is hard to implement. I will use sth else instead. In fact I miss C++ and my `#define`s.

File: src/syntax/concrete.rs (hunk 1 above)
Comment (HoshinoTented) [reply src/syntax/concrete.rs]: Ummmm, what is Visitor?

File: src/diag.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        })?;
+        Ok(())
+    }
+}
+
+pub struct Diagnostic {
+    pub level: DiagLevel,
+    pub info: String,
+    pub source_range: SourceRange,
+    pub fix_hint: Option<String>,
```
Comment (HoshinoTented) [inline src/diag.rs]: Should we wrap `fix_hint` and `fix_hint_source_range` into a specific structure?

File: src/syntax/lex.rs [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+                break;
+            }
+        }
+        self.token_stream.push(Token::new(
+            TokenKind::LitNumber(num),
+            SourceRange::new(self.line, start_col, self.column)
+        ));
+    }
+
+    fn lex_char(&mut self) {
```
Comment (HoshinoTented) [inline src/syntax/lex.rs:118]: the escape character of HNM is '?', such as "?n", "?t", "??"

File: src/syntax/concrete.rs (hunk 1 above)
Comment (chuigda) [reply src/syntax/concrete.rs]: Finally I gave up visitor pattern and I finally choose to use `match`.

File: src/diag.rs (hunk 2 above)
Comment (chuigda) [reply src/diag.rs]: да. I will do that.

File: src/syntax/lex.rs (hunk 3 above)
Comment (chuigda) [reply src/syntax/lex.rs:118]: fixed in new commit.

**Comment by chuigda** [comment]
Look at these fancy trees~

File: src/syntax/concrete/stmt.rs [hunk 4]
```diff
@@ -1,12 +1,38 @@
+use crate::syntax::concrete::{Expr, Decl};
+use crate::syntax::token::SourceRange;
+
 pub enum Stmt {
+    ExprStmt(ExprStmt),
+    DeclStmt(DeclStmt),
+    BlockStmt(BlockStmt)
```
Comment (HoshinoTented) [inline src/syntax/concrete/stmt.rs]: `Block` is also an `Expr`

Comment (chuigda) [reply src/syntax/concrete/stmt.rs]: да.

**Comment by chuigda** [comment]
I'll remake one. I have better design (maybe) now.

**Pull request closed without merging**